File: Java/ExtensibleCommands/src/test/java/org/extensiblecommands/ConditionalCommandTest.java
package org.extensiblecommands;

import org.junit.*;
import org.junit.rules.TestName;

public class ConditionalCommandTest {
    @Rule
    public TestName name = new TestName();

    @BeforeClass
    public static void setUpClass() {
        Setup.InitLog();
    }

    @Before
    public void setUpTest() {
        Logger.log(Logger.LogLevel.Info,
                "----------------------------------------------------------------------------------------------------------");
        Logger.log(Logger.LogLevel.Info,
                String.format("Starting Test %s:%s", this.getClass().getName(), name.getMethodName()));
    }

    @Test
    public void constructionTest() {
        var command = new ConditionalCommand(() -> true, SimpleCommand.NullCommand,
                SimpleCommand.NullCommand);
        Assert.assertEquals(command.getTrueCommand(), SimpleCommand.NullCommand);
        Assert.assertEquals(command.getFalseCommand(), SimpleCommand.NullCommand);
        Assert.assertEquals(command.getName(), "Conditional");

        command = new ConditionalCommand(() -> true, SimpleCommand.NullCommand,
                SimpleCommand.NullCommand, "MyCommand");
        Assert.assertEquals(command.getTrueCommand(), SimpleCommand.NullCommand);
        Assert.assertEquals(command.getFalseCommand(), SimpleCommand.NullCommand);
        Assert.assertEquals(command.getName(), "MyCommand");

        // Malformed cases
        boolean exceptionCaught = false;
        try {
            new ConditionalCommand(() -> true, null, null, "");
        }
        catch (Exception e) {
            if (e.getMessage().contains("is NULL"))
                exceptionCaught = true;
        }
        Assert.assertTrue(exceptionCaught);

        exceptionCaught = false;
        try {
            new ConditionalCommand(() -> true, SimpleCommand.NullCommand, null, "");
        }
        catch (Exception e) {
            if (e.getMessage().contains("is NULL"))
                exceptionCaught = true;
        }
        Assert.assertTrue(exceptionCaught);

        exceptionCaught = false;
        try {
            command = new ConditionalCommand(() -> true, SimpleCommand.NullCommand, SimpleCommand.NullCommand, "");
            command.run();
        }
        catch (Exception e) {
            if (e.getMessage().contains("is NULL"))
                exceptionCaught = true;
        }
        Assert.assertFalse(exceptionCaught);
    }

    @Test
    public void trueCommandTest() throws Exception {
        var conditionalCommand = createConditionalCommand(true);

        Setup.runAndWaitForNormalCompletion(conditionalCommand);

        Assert.assertEquals(State.Completed, conditionalCommand.getState());
        Assert.assertEquals(State.Completed, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getFalseCommand().getState());
    }

    @Test
    public void falseCommandTest() throws Exception {
        var conditionalCommand = createConditionalCommand(false);

        Setup.runAndWaitForNormalCompletion(conditionalCommand);

        Assert.assertEquals(State.Completed, conditionalCommand.getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Completed, conditionalCommand.getFalseCommand().getState());
    }

    @Test
    public void runTrueCommandErrorTest() throws Exception {
        var trueCommand = new SimpleCommand(() -> { throw new ExtensibleCommandsException(Setup.TestErrorCode, Setup.TestErrorDescription); }, "True");
        var falseCommand = new SimpleCommand(() -> { }, "False");
        var conditionalCommand = new ConditionalCommand(() -> true, trueCommand, falseCommand, "Conditional");

        Setup.runAndWaitForFailure(conditionalCommand);

        Assert.assertEquals(State.Failed, conditionalCommand.getState());
        Assert.assertEquals(State.Failed, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getFalseCommand().getState());

        Assert.assertEquals(Setup.TestErrorCode, conditionalCommand.getException().getId());
        Assert.assertEquals(Setup.TestErrorDescription, conditionalCommand.getException().getText());
    }

    @Test
    public void runFalseCommandErrorTest() throws Exception {
        var trueCommand = new SimpleCommand(() -> { }, "True");
        var falseCommand = new SimpleCommand(() -> { throw new ExtensibleCommandsException(Setup.TestErrorCode, Setup.TestErrorDescription); }, "False");

        var conditionalCommand = new ConditionalCommand(() -> false, trueCommand, falseCommand, "Conditional");

        Setup.runAndWaitForFailure(conditionalCommand);

        Assert.assertEquals(State.Failed, conditionalCommand.getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Failed, conditionalCommand.getFalseCommand().getState());

        Assert.assertEquals(Setup.TestErrorCode, conditionalCommand.getException().getId());
        Assert.assertEquals(Setup.TestErrorDescription, conditionalCommand.getException().getText());
    }

    @Test
    public void abortTrueTest() throws Exception {

        var conditionalCommand = createConditionalBranchStopAbortCommand(true, false);

        Setup.runAndWaitForAbort(conditionalCommand);

        Assert.assertEquals(State.Aborted, conditionalCommand.getState());
        Assert.assertEquals(State.Aborted, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getFalseCommand().getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(1).getState());
        Assert.assertEquals(State.Idle, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(2).getState());
    }

    @Test
    public void pauseResumeTrueTest() throws Exception {
        var conditionalCommand = createConditionalBranchStopAbortCommand(true, true);

        ExecutionDelegate assertAfterPause = () -> {
            Assert.assertEquals(State.Executing, conditionalCommand.getState());
            Assert.assertEquals(State.Executing, conditionalCommand.getTrueCommand().getState());
            Assert.assertEquals(State.Idle, conditionalCommand.getFalseCommand().getState());
            Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(0).getState());
            Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(1).getState());
            Assert.assertEquals(State.Idle, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(2).getState());
        };

        Setup.pauseAndResume(conditionalCommand, assertAfterPause);

        Assert.assertEquals(State.Completed, conditionalCommand.getState());
        Assert.assertEquals(State.Completed, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getFalseCommand().getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(1).getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(2).getState());
    }

    @Test
    public void pauseAbortTrueTest() throws Exception {
        var conditionalCommand = createConditionalBranchStopAbortCommand(true, true);

        ExecutionDelegate assertAfterPause = () -> {
            Assert.assertEquals(State.Executing, conditionalCommand.getState());
            Assert.assertEquals(State.Executing, conditionalCommand.getTrueCommand().getState());
            Assert.assertEquals(State.Idle, conditionalCommand.getFalseCommand().getState());
            Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(0).getState());
            Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(1).getState());
            Assert.assertEquals(State.Idle, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(2).getState());
        };

        Setup.pauseAndAbort(conditionalCommand, assertAfterPause);

        Assert.assertEquals(State.Aborted, conditionalCommand.getState());
        Assert.assertEquals(State.Aborted, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getFalseCommand().getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(1).getState());
        Assert.assertEquals(State.Idle, ((SequentialCommand)conditionalCommand.getTrueCommand()).getSubCommand(2).getState());
    }

    @Test
    public void abortFalseTest() throws Exception {

        var conditionalCommand = createConditionalBranchStopAbortCommand(false, false);

        Setup.runAndWaitForAbort(conditionalCommand);

        Assert.assertEquals(State.Aborted, conditionalCommand.getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Aborted, conditionalCommand.getFalseCommand().getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(1).getState());
        Assert.assertEquals(State.Idle, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(2).getState());
    }

    @Test
    public void pauseResumeFalseTest() throws Exception {
        var conditionalCommand = createConditionalBranchStopAbortCommand(false, true);

        ExecutionDelegate assertAfterPause = () -> {
            Assert.assertEquals(State.Executing, conditionalCommand.getState());
            Assert.assertEquals(State.Idle, conditionalCommand.getTrueCommand().getState());
            Assert.assertEquals(State.Executing, conditionalCommand.getFalseCommand().getState());
            Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(0).getState());
            Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(1).getState());
            Assert.assertEquals(State.Idle, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(2).getState());
        };

        Setup.pauseAndResume(conditionalCommand, assertAfterPause);

        Assert.assertEquals(State.Completed, conditionalCommand.getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Completed, conditionalCommand.getFalseCommand().getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(1).getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(2).getState());
    }

    @Test
    public void pauseAbortFalseTest() throws Exception {
        var conditionalCommand = createConditionalBranchStopAbortCommand(false, true);

        ExecutionDelegate assertAfterPause = () -> {
            Assert.assertEquals(State.Executing, conditionalCommand.getState());
            Assert.assertEquals(State.Idle, conditionalCommand.getTrueCommand().getState());
            Assert.assertEquals(State.Executing, conditionalCommand.getFalseCommand().getState());
            Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(0).getState());
            Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(1).getState());
            Assert.assertEquals(State.Idle, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(2).getState());
        };

        Setup.pauseAndAbort(conditionalCommand, assertAfterPause);

        Assert.assertEquals(State.Aborted, conditionalCommand.getState());
        Assert.assertEquals(State.Idle, conditionalCommand.getTrueCommand().getState());
        Assert.assertEquals(State.Aborted, conditionalCommand.getFalseCommand().getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(0).getState());
        Assert.assertEquals(State.Completed, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(1).getState());
        Assert.assertEquals(State.Idle, ((SequentialCommand)conditionalCommand.getFalseCommand()).getSubCommand(2).getState());    }

    @Test
    public void retrieveSubCommands() {
        var command1 = new SequentialCommand("Core")
            .add(new SimpleCommand("1"))
            .add(new SimpleCommand("2"))
            .add(new SimpleCommand("3"));

        var command2 = new SequentialCommand("True")
            .add(new SimpleCommand("T-1"))
            .add(new SimpleCommand("T-2"))
            .add(new SimpleCommand("T-3"))
            .add(new SimpleCommand("T-4"))
            .add(new SimpleCommand("T-5"));

        var command = new ConditionalCommand(() -> true, command1, command2, "Main");

        Assert.assertEquals(3, command1.getDescendants().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(5, command2.getDescendants().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(10, command.getDescendants().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(3, command1.getChildren().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(5, command2.getChildren().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(2, command.getChildren().spliterator().getExactSizeIfKnown());
    }

    @Test
    public void runInputOutputTest() throws Exception {
        var command = new SimpleCommandIO<String, Integer>(input -> input.length(), "Test");

        var conditionalCommand = new ConditionalCommand(() -> true,
                command, SimpleCommand.NullCommand, "");

        command.setInput("input");
        conditionalCommand.run();
    }
    
    //----------------------------------------------------------------------------------------------------------------------

    private ConditionalCommand createConditionalCommand(boolean flag) {
        var trueCommand = new SimpleCommand(() -> { }, "True");
        var falseCommand = new SimpleCommand(() -> { }, "False");
        return new ConditionalCommand(() -> flag, trueCommand, falseCommand, "Conditional");
    }

    private ConditionalCommand createConditionalBranchStopAbortCommand(boolean branch, boolean pause) {
        Command trueCommand, falseCommand;

        SequentialCommand seqCommand;
        if (branch) {
            seqCommand = new SequentialCommand("True");
            trueCommand = seqCommand;
            falseCommand = new SimpleCommand(() -> { }, "False");
        }
        else {
            seqCommand = new SequentialCommand("False");
            trueCommand = new SimpleCommand(() -> { }, "True");
            falseCommand = seqCommand;
        }

        var conditionalCommand = new ConditionalCommand(() -> branch, trueCommand, falseCommand, "Conditional");

        seqCommand.add(new SimpleCommand(() -> { }, "S1"));

        if (pause)
            seqCommand.add(new SimpleCommand(() -> conditionalCommand.pause(), "S2-Pause"));
        else
            seqCommand.add(new SimpleCommand(() -> conditionalCommand.abort(), "S2-abort"));
        seqCommand.add(new SimpleCommand(() -> { }, "S3"));

        return conditionalCommand;
    }
}

File: CSharp/ExtensibleCommands/ExtensibleCommandsUnitTests/ConditionalCommandTest.cs
using System;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;

using ExtensibleCommands;

namespace ExtensibleCommandsUnitTest
{
    /// <summary>
    /// Summary description for UnitTest1
    /// </summary>
    [TestClass]
    public class ConditionalCommandTest
    {
    
        private TestContext testContextInstance;

        /// <summary>
        ///Gets or sets the test context which provides
        ///information about and functionality for the current test run.
        ///</summary>
        public TestContext TestContext
        {
            get
            {
                return testContextInstance;
            }
            set
            {
                testContextInstance = value;
            }
        }

        #region Additional test attributes
        // 
        //You can use the following additional attributes as you write your tests:
        //
        //Use ClassInitialize to run code before running the first test in the class
        [ClassInitialize()]
        public static void MyClassInitialize(TestContext testContext)
        {
        }
        //
        //Use ClassCleanup to run code after all tests in a class have run
        //[ClassCleanup()]
        //public static void MyClassCleanup()
        //{
        //}
        //
        //Use TestInitialize to run code before running each test
        [TestInitialize()]
        public void MyTestInitialize()
        {
            Logger.Log(Logger.LogLevel.Info,
                "----------------------------------------------------------------------------------------------------------");
            Logger.Log(Logger.LogLevel.Info,
                string.Format("Starting Test {0}:{1}", GetType().Name, testContextInstance.TestName));
        }
        //
        //Use TestCleanup to run code after each test has run
        //[TestCleanup()]
        //public void MyTestCleanup()
        //{
        //}
        //
        #endregion

        [TestMethod()]
        public void ConstructionTest()
        {
            var command = new ConditionalCommand(() => true, SimpleCommand.NullCommand,
                SimpleCommand.NullCommand);
            Assert.AreEqual(command.TrueCommand, SimpleCommand.NullCommand);
            Assert.AreEqual(command.FalseCommand, SimpleCommand.NullCommand);
            Assert.AreEqual(command.Name, "Conditional");

            command = new ConditionalCommand(() => true, SimpleCommand.NullCommand,
                SimpleCommand.NullCommand, "MyCommand");
            Assert.AreEqual(command.TrueCommand, SimpleCommand.NullCommand);
            Assert.AreEqual(command.FalseCommand, SimpleCommand.NullCommand);
            Assert.AreEqual(command.Name, "MyCommand");

            // Malformed cases
            bool exceptionCaught = false;
            try
            {
                new ConditionalCommand(() => true, null, null, "");
            }
            catch (Exception e)
            {
                if (e.Message.Contains("is NULL"))
                    exceptionCaught = true;
            }
            Assert.IsTrue(exceptionCaught);

            exceptionCaught = false;
            try
            {
                new ConditionalCommand(() => true, SimpleCommand.NullCommand, null, "");
            }
            catch (Exception e)
            {
                if (e.Message.Contains("is NULL"))
                    exceptionCaught = true;
            }
            Assert.IsTrue(exceptionCaught);

            exceptionCaught = false;
            try
            {
                command = new ConditionalCommand(() => true, SimpleCommand.NullCommand, SimpleCommand.NullCommand, "");
                command.Run();
            }
            catch (Exception e)
            {
                if (e.Message.Contains("is NULL"))
                    exceptionCaught = true;
            }
            Assert.IsFalse(exceptionCaught);
        }

        [TestMethod()]
        public void TrueCommandTest()
        {
            var conditionalCommand = CreateConditionalCommand(true);

            Setup.RunAndWaitForNormalCompletion(conditionalCommand);

            Assert.AreEqual(State.Completed, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Completed, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.FalseCommand.CurrentState);
        }

        [TestMethod()]
        public void FalseCommandTest()
        {
            var conditionalCommand = CreateConditionalCommand(false);

            Setup.RunAndWaitForNormalCompletion(conditionalCommand);

            Assert.AreEqual(State.Completed, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Completed, conditionalCommand.FalseCommand.CurrentState);
        }

        [TestMethod()]
        public void RunTrueCommandErrorTest()
        {
            var trueCommand = new SimpleCommand(() => { throw new ExtensibleCommandsException(Setup.TestErrorCode, Setup.TestErrorDescription); }, "True");
            var falseCommand = new SimpleCommand(() => { }, "False");
            var conditionalCommand = new ConditionalCommand(() => true, trueCommand, falseCommand, "Conditional");

            Setup.RunAndWaitForFailure(conditionalCommand);

            Assert.AreEqual(State.Failed, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Failed, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.FalseCommand.CurrentState);

            Assert.AreEqual(Setup.TestErrorCode, conditionalCommand.Exception.ID);
            Assert.AreEqual(Setup.TestErrorDescription, conditionalCommand.Exception.Text);
        }

        [TestMethod()]
        public void RunFalseCommandErrorTest()
        {
            var trueCommand = new SimpleCommand(() => { }, "True");
            var falseCommand = new SimpleCommand(() => { throw new ExtensibleCommandsException(Setup.TestErrorCode, Setup.TestErrorDescription); }, "False");
            var conditionalCommand = new ConditionalCommand(() => false, trueCommand, falseCommand, "Conditional");

            Setup.RunAndWaitForFailure(conditionalCommand);

            Assert.AreEqual(State.Failed, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Failed, conditionalCommand.FalseCommand.CurrentState);

            Assert.AreEqual(Setup.TestErrorCode, conditionalCommand.Exception.ID);
            Assert.AreEqual(Setup.TestErrorDescription, conditionalCommand.Exception.Text);
        }

        [TestMethod()]
        public void AbortTrueTest()
        {
            var conditionalCommand = CreateConditionalBranchStopAbortCommand(true, false);

            Setup.RunAndWaitForAbort(conditionalCommand);

            Assert.AreEqual(State.Aborted, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Aborted, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.FalseCommand.CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Idle, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(2).CurrentState);
        }

        [TestMethod()]
        public void PauseResumeTrueTest()
        {
            var conditionalCommand = CreateConditionalBranchStopAbortCommand(true, true);

            var assertAfterPause = new System.Action(() =>
            {
                Assert.AreEqual(State.Executing, conditionalCommand.CurrentState);
                Assert.AreEqual(State.Executing, conditionalCommand.TrueCommand.CurrentState);
                Assert.AreEqual(State.Idle, conditionalCommand.FalseCommand.CurrentState);
                Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(0).CurrentState);
                Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(1).CurrentState);
                Assert.AreEqual(State.Idle, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(2).CurrentState);
            });

            Setup.PauseAndResume(conditionalCommand, assertAfterPause);

            Assert.AreEqual(State.Completed, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Completed, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.FalseCommand.CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(2).CurrentState);
        }

        [TestMethod()]
        public void PauseAbortTrueTest()
        {
            var conditionalCommand = CreateConditionalBranchStopAbortCommand(true, true);

            var assertAfterPause = new System.Action(() =>
            {
                Assert.AreEqual(State.Executing, conditionalCommand.CurrentState);
                Assert.AreEqual(State.Executing, conditionalCommand.TrueCommand.CurrentState);
                Assert.AreEqual(State.Idle, conditionalCommand.FalseCommand.CurrentState);
                Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(0).CurrentState);
                Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(1).CurrentState);
                Assert.AreEqual(State.Idle, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(2).CurrentState);
            });

            Setup.PauseAndAbort(conditionalCommand, assertAfterPause);

            Assert.AreEqual(State.Aborted, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Aborted, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.FalseCommand.CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Idle, (conditionalCommand.TrueCommand as SequentialCommand).GetSubCommand(2).CurrentState);
        }

        [TestMethod()]
        public void AbortFalseTest()
        {
            var conditionalCommand = CreateConditionalBranchStopAbortCommand(false, false);

            Setup.RunAndWaitForAbort(conditionalCommand);

            Assert.AreEqual(State.Aborted, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Aborted, conditionalCommand.FalseCommand.CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Idle, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(2).CurrentState);
        }

        [TestMethod()]
        public void PauseResumeFalseTest()
        {
            var conditionalCommand = CreateConditionalBranchStopAbortCommand(false, true);

            var assertAfterPause = new System.Action(() =>
            {
                Assert.AreEqual(State.Executing, conditionalCommand.CurrentState);
                Assert.AreEqual(State.Idle, conditionalCommand.TrueCommand.CurrentState);
                Assert.AreEqual(State.Executing, conditionalCommand.FalseCommand.CurrentState);
                Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(0).CurrentState);
                Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(1).CurrentState);
                Assert.AreEqual(State.Idle, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(2).CurrentState);
            });

            Setup.PauseAndResume(conditionalCommand, assertAfterPause);

            Assert.AreEqual(State.Completed, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Completed, conditionalCommand.FalseCommand.CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(2).CurrentState);
        }

        [TestMethod()]
        public void PauseAbortFalseTest()
        {
            var conditionalCommand = CreateConditionalBranchStopAbortCommand(false, true);

            var assertAfterPause = new System.Action(() =>
            {
                Assert.AreEqual(State.Executing, conditionalCommand.CurrentState);
                Assert.AreEqual(State.Idle, conditionalCommand.TrueCommand.CurrentState);
                Assert.AreEqual(State.Executing, conditionalCommand.FalseCommand.CurrentState);
                Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(0).CurrentState);
                Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(1).CurrentState);
                Assert.AreEqual(State.Idle, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(2).CurrentState);
            });

            Setup.PauseAndAbort(conditionalCommand, assertAfterPause);

            Assert.AreEqual(State.Aborted, conditionalCommand.CurrentState);
            Assert.AreEqual(State.Idle, conditionalCommand.TrueCommand.CurrentState);
            Assert.AreEqual(State.Aborted, conditionalCommand.FalseCommand.CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(0).CurrentState);
            Assert.AreEqual(State.Completed, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(1).CurrentState);
            Assert.AreEqual(State.Idle, (conditionalCommand.FalseCommand as SequentialCommand).GetSubCommand(2).CurrentState);
        }

        [TestMethod()]
        public void RetrieveSubCommandsTest()
        {
            var command1 = new SequentialCommand("Core")
                .Add(new SimpleCommand("1"))
                .Add(new SimpleCommand("2"))
                .Add(new SimpleCommand("3"));

            var command2 = new SequentialCommand("True")
                .Add(new SimpleCommand("T-1"))
                .Add(new SimpleCommand("T-2"))
                .Add(new SimpleCommand("T-3"))
                .Add(new SimpleCommand("T-4"))
                .Add(new SimpleCommand("T-5"));

            var command = new ConditionalCommand(() => true, command1, command2, "Main");

            Assert.AreEqual(3, command1.Descendants.Count());
            Assert.AreEqual(5, command2.Descendants.Count());
            Assert.AreEqual(10, command.Descendants.Count());
            Assert.AreEqual(3, command1.Children.Count());
            Assert.AreEqual(5, command2.Children.Count());
            Assert.AreEqual(2, command.Children.Count());
        }

        [TestMethod()]
        public void RunInputOutputTest()
        {
            var command = new SimpleCommandIO<string, int>(input => input.Length, "Test");
            
            var conditionalCommand = new ConditionalCommand(() => true,
                command, SimpleCommand.NullCommand, "");

            command.Input = "input";
            conditionalCommand.Run();
        }

        //----------------------------------------------------------------------------------------------------------------------

        private ConditionalCommand CreateConditionalCommand(bool flag)
        {
            var trueCommand = new SimpleCommand(() => { }, "True");
            var falseCommand = new SimpleCommand(() => { }, "False");
            return new ConditionalCommand(() => flag, trueCommand, falseCommand, "Conditional");
        }

        private ConditionalCommand CreateConditionalBranchStopAbortCommand(bool branch, bool pause)
        {
            ICommand trueCommand, falseCommand;

            SequentialCommand seqCommand;
            if (branch)
            {
                seqCommand = new SequentialCommand("True");
                trueCommand = seqCommand;
                falseCommand = new SimpleCommand(() => { }, "False");
            }
            else
            {
                seqCommand = new SequentialCommand("False");
                trueCommand = new SimpleCommand(() => { }, "True");
                falseCommand = seqCommand;
            }

            var conditionalCommand = new ConditionalCommand(() => branch, trueCommand, falseCommand, "Conditional");

            seqCommand.Add(new SimpleCommand(() => { }, "S1"));

            if (pause)
                seqCommand.Add(new SimpleCommand(() => conditionalCommand.Pause(), "S2-Pause"));
            else
                seqCommand.Add(new SimpleCommand(() => conditionalCommand.Abort(), "S2-Abort"));
            seqCommand.Add(new SimpleCommand(() => { }, "S3"));

            return conditionalCommand;
        }
    }
}
